server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java

```diff
@@ -99,12 +101,24 @@ public class CoordinationDiagnosticsService implements ClusterStateListener {
     // Non-private for testing
     volatile ConcurrentMap<DiscoveryNode, ClusterFormationStateOrException> clusterFormationResponses = null;
 
+    /*
+     * This is a reference to the task that is periodically reaching out to a master eligible node to get its CoordinationDiagnosticsResult
+     * for diagnosis. It is null when no polling is occurring.
+     * The field is accessed (reads/writes) from multiple threads, and is also reassigned on multiple threads.
+     */
+    volatile AtomicReference<Scheduler.Cancellable> remoteStableMasterHealthIndicatorTask = null;
```

Comment:
Does this need to be an `AtomicReference` or a `volatile Cancellable` would suffice?

Reply:
We potentially assign this multiple times on the same polling "stack". So I think you're right that it doesn't necessarily have to be atomic, but it has to be some reference because we need to access it from the cancel method. That is, if I only had a Cancellable here, I couldn't reassign that same Cancellable reference to a new Cancellable and still have access to it in the cancel method. I can't just assign it to the object-level variable here because there might be more than one polling stack running at the same time.